File: Semana_Epidemiologica/SE_fetcher.py
from playwright.sync_api import sync_playwright, TimeoutError
import yaml
import time

def main():
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)  # Run in non-headless mode for debugging
        page = browser.new_page()
        page.goto("https://www.gov.br/saude/pt-br/assuntos/saude-de-a-a-z/a/aedes-aegypti/monitoramento-das-arboviroses")
        
        if page.locator("button.btn-accept").is_visible(timeout=5000):
            page.locator("button.btn-accept").click()
            print("Cookies accepted")
        
        page.wait_for_load_state("networkidle")
        print("Page loaded")

        try:
            # Access the iframe
            frame_locator = page.frame_locator("iframe[title='Sala Nacional de Arboviroses - SNA']")
            frame = frame_locator
            frame.locator("body").wait_for(state="visible", timeout=15000)
            print("Iframe accessed")

            # Select the Dengue panel
            dengue_element = frame.get_by_role("group").filter(has_text="Dengue").locator("path").first
            dengue_element.wait_for(state="visible", timeout=10000)
            dengue_element.scroll_into_view_if_needed()
            dengue_element.click(delay=100)
            print("Dengue panel selected")

            # Target the SEM_PRI_SE dropdown
            dropdown = frame.locator("div.slicer-dropdown-menu[aria-label='SEM_PRI_SE']")
            dropdown.wait_for(state="visible", timeout=10000)
            dropdown.click()
            print("SEM_PRI_SE dropdown opened")

            # Immediately select "Select all"
            select_all_locator = page.locator("iframe[title='Sala Nacional de Arboviroses - SNA']").content_frame.get_by_text("Select all")
            select_all_locator.wait_for(state="visible", timeout=10000)
            select_all_locator.click()
            print("Checked 'Select all'")
            time.sleep(1)  # Brief wait for UI to settle

            # Function to get all menu items in SEM_PRI_SE
            def get_menu_items():
                return frame.locator("div.slicer-dropdown-popup div.slicerItemContainer").all()

            # Scroll through the dropdown to find the last element
            previous_count = 0
            menu_items = get_menu_items()
            max_iterations = 10
            iteration = 0
            last_element_title = None

            while iteration < max_iterations:
                current_count = len(menu_items)
                if current_count == previous_count and current_count > 0:
                    print(f"No new items found. Total items: {current_count}")
                    last_element_title = menu_items[-1].get_attribute("title").strip()
                    print(f"Last epidemiological week identified: {last_element_title}")
                    break

                if menu_items:
                    last_item = menu_items[-1]
                    last_item.scroll_into_view_if_needed()
                    print(f"Scrolled to item {current_count}: {last_item.text_content()}")
                    time.sleep(1)  # Wait for lazy loading

                menu_items = get_menu_items()
                previous_count = current_count
                iteration += 1

            # Fetch data with "Select all" checked
            if last_element_title:
                # Wait for the UI to update with all data
                time.sleep(3)
                frame.locator("svg.card").first.wait_for(state="visible", timeout=15000)

                sem_data = {
                    "Last_Epidemiological_Week": last_element_title,
                    "All_Semanas": {}
                }
                svg_cards = frame.locator("svg.card").all()

                for card in svg_cards:
                    aria_label = card.get_attribute("aria-label")
                    value = card.locator("text.value tspan").text_content()
                    if aria_label and value:
                        sem_data["All_Semanas"][aria_label] = value

                # Save data to YAML
                with open("Semana_Epidemiologica/output/SE-Y.yaml", "w", encoding="utf-8") as f:
                    yaml.dump(sem_data, f, allow_unicode=True, default_flow_style=False)
                print("Data saved to dengue_sem_data.yaml")

            else:
                print("No last element identified within iteration limit.")
                page.pause()

        except TimeoutError as e:
            print(f"Timeout error: {e}")
            page.pause()
        except Exception as e:
            print(f"An error occurred: {e}")
            page.pause()
        finally:
            browser.close()

if __name__ == "__main__":
    main()

File: data_fetching_pipeline/SEFetcher.js
const fs = require('fs');
const yaml = require('js-yaml');
const path = require('path');
const { navigateToDengue, generateDatedFilename, extractCardsData } = require('./utils');

async function SEFetcher() {
    let browser;
    try {
        console.log("Navigating to Dengue panel...");
        const { browser: launchedBrowser, page } = await navigateToDengue();
        browser = launchedBrowser;
        page.setDefaultTimeout(25000);

        const dropdownButton = page.locator("div.slicer-dropdown-menu[aria-label='SEM_PRI_SE']");
        await dropdownButton.waitFor({ state: "visible", timeout: 15000 });
        await dropdownButton.click();
        console.log("SEM_PRI_SE dropdown opened.");

        let popup = page.locator("div.slicer-dropdown-popup.focused");
        try {
            await popup.waitFor({ state: "visible", timeout: 10000 });
        } catch {
            console.log("Could not find focused popup. Trying generic visible...");
            popup = page.locator("div.slicer-dropdown-popup:visible").first();
            await popup.waitFor({ state: "visible", timeout: 5000 });
        }

        const container = popup.locator(".slicerBody");
        await container.waitFor({ state: "visible", timeout: 10000 });
        console.log("Items container located.");

        const selectAll = popup.getByText("Select all", { exact: true });
        await selectAll.waitFor({ state: "visible", timeout: 10000 });
        const isChecked = await selectAll.evaluate(node =>
            node.checked || node.getAttribute("aria-checked") === "true" || node.parentElement.classList.contains("selected")
        );
        if (!isChecked) {
            await selectAll.click();
            console.log("Checked 'Select all'.");
            await page.waitForTimeout(1000);
        } else {
            console.log("'Select all' was already checked.");
        }

        console.log("Starting infinite scroll...");
        let lastWeek = null;
        let previousTitle = null;
        let attempts = 0;
        const maxAttempts = 50;

        while (attempts < maxAttempts) {
            attempts++;
            console.log(`Scroll attempt ${attempts}/${maxAttempts}`);

            const items = await container.locator("div.slicerItemContainer").all();
            const numericItems = [];

            for (const item of items) {
                const title = await item.getAttribute("title");
                if (title && /^\d+$/.test(title.trim())) {
                    numericItems.push({ element: item, title: title.trim() });
                }
            }

            if (numericItems.length === 0) {
                if (attempts <= 3) await page.waitForTimeout(1000);
                else {
                    console.warn("No numeric items found after several attempts.");
                    break;
                }
                continue;
            }

            const lastItem = numericItems[numericItems.length - 1];
            const currentTitle = lastItem.title;
            console.log(`Current last visible week: ${currentTitle}`);

            if (attempts > 1 && currentTitle === previousTitle) {
                console.log("End of list detected.");
                lastWeek = currentTitle;
                break;
            }

            previousTitle = currentTitle;

            try {
                await lastItem.element.scrollIntoViewIfNeeded({ timeout: 5000 });
                await page.waitForTimeout(1500);
            } catch (e) {
                console.error(`Error during scroll: ${e}`);
                lastWeek = previousTitle;
                break;
            }
        }

        if (!lastWeek && previousTitle) {
            console.log(`Using last known week: ${previousTitle}`);
            lastWeek = previousTitle;
        }

        if (!lastWeek) {
            console.warn("Could not determine last week.");
            return;
        }

        console.log(`Fetching data for week ${lastWeek}...`);
        await page.waitForTimeout(5000);
        await page.locator("svg.card").first().waitFor({ state: "visible", timeout: 30000 });

        const cardData = await extractCardsData(page, { includeLetalidade: true });
        const semData = {
            Last_Epidemiological_Week: lastWeek,
            All_Semanas_Data: cardData
        };

        const dateStr = new Date().toISOString().split('T')[0].replace(/-/g, '_');
        const outputDir = path.join(__dirname, '..', 'output', `dengue_${dateStr}`);
        fs.mkdirSync(outputDir, { recursive: true });
        const filePath = generateDatedFilename(`SE-Y-${lastWeek}`, "yaml", outputDir);
        fs.writeFileSync(filePath, yaml.dump(semData, { noRefs: true, sortKeys: false }), 'utf8');
        console.log(`✅ Data saved to ${filePath}`);
    } catch (err) {
        console.error(`❌ Error in SEFetcher: ${err.message}`);
        throw err;
    } finally {
        if (browser) {
            console.log("Closing browser.");
            await browser.close();
        }
    }
}

module.exports = { SEFetcher };

if (require.main === module) {
    SEFetcher().catch(console.error);
}